File: listing_service.py
from sqlalchemy import create_engine, text
import os

# Lokasi database listing
base_dir = '/home/ubuntu/botwa'
DB_PATH = os.path.join(base_dir, 'listing.db')

def get_listing_caption(plu):
    """
    Menerima PLU, mencari di database listing.db, 
    dan mengembalikan string caption lengkap untuk bot.
    """
    if not os.path.exists(DB_PATH):
        return "⚠️ Error: Database listing.db belum dibuat."

    try:
        engine = create_engine(f'sqlite:///{DB_PATH}')
        with engine.connect() as conn:
            # Query data berdasarkan PLU
            query = text("SELECT * FROM listings WHERE PLU = :plu")
            result = conn.execute(query, {"plu": str(plu).strip()}).fetchone()
            
            if result:
                # Mapping hasil query (sesuai urutan kolom di import_all_listings.py)
                # 0=PLU, 1=NAMA BARANG, 2=RAK, 3=SHELFING, 4=BARIS, 5=RETUR
                nama = result[1]
                rak = result[2]
                shelf = result[3]
                baris = result[4]
                retur = result[5]
                
                # Format Caption sesuai permintaan (di bawah gambar)
                caption = f"*{nama}*\n*Rak:* {rak} | *Shelf:* {shelf} | *Baris:* {baris}\n*Retur:* {retur}"
                return caption
            else:
                return None # Data tidak ditemukan
                
    except Exception as e:
        return f"⚠️ Error System: {e}"

File: listing_service.js
const sqlite3 = require('sqlite3').verbose();
const fs = require('fs');

// Path ke database listing.db (Pastikan path ini benar dan sama dengan script Python)
const DB_PATH = '/home/oem/Documents/bail/listing.db';

/**
 * Mengambil caption lengkap (Nama, Rak, Shelf, Baris, Retur) berdasarkan PLU.
 * @param {string|number} plu 
 * @returns {Promise<string|null>} Caption string atau null jika tidak ditemukan.
 */
function getListingCaption(plu) {
    return new Promise((resolve, reject) => {
        if (!fs.existsSync(DB_PATH)) {
            // console.error(`❌ Database tidak ditemukan di: ${DB_PATH}`);
            // Resolve null agar bot tidak error fatal, hanya tidak ada caption
            return resolve(null);
        }

        const db = new sqlite3.Database(DB_PATH, sqlite3.OPEN_READONLY, (err) => {
            if (err) {
                console.error("❌ Gagal membuka database listing:", err.message);
                return resolve(null);
            }
        });

        // Query mencari PLU
        const sql = `SELECT * FROM listings WHERE PLU = ?`;
        const pluString = String(plu).trim();

        db.get(sql, [pluString], (err, row) => {
            db.close(); // Tutup koneksi setelah query selesai

            if (err) {
                console.error("❌ Error saat query PLU:", err.message);
                return resolve(null);
            }

            if (row) {
                // Ambil data kolom (sesuai nama kolom di Python script)
                const nama = row['NAMA BARANG'];
                const rak = row['RAK'];
                const shelf = row['SHELFING'];
                const baris = row['BARIS'];
                const retur = row['RETUR'];

                // Format Caption
                const caption = `${nama}\nRak: ${rak} | Shelf: ${shelf} | Baris: ${baris}\nRetur: ${retur}`;
                resolve(caption);
            } else {
                // Data tidak ditemukan
                resolve(null);
            }
        });
    });
}

module.exports = { getListingCaption };